ext/socket/lib/socket.rb

```diff
@@ -623,45 +684,131 @@ def accept_nonblock(exception: true)
   def self.tcp(host, port, local_host = nil, local_port = nil, connect_timeout: nil, resolv_timeout: nil) # :yield: socket
     last_error = nil
     ret = nil
+    attempt_v6 = true
+    attempt_v4 = true
+    start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+
+    if connect_timeout
+      raise ArgumentError, "connect_timeout must be Numeric" unless connect_timeout.is_a?(Numeric)
+      raise ArgumentError, "connect_timeout must not be negative" if connect_timeout.negative?
+    end
 
     local_addr_list = nil
-    if local_host != nil || local_port != nil
+    if local_host && local_port
       local_addr_list = Addrinfo.getaddrinfo(local_host, local_port, nil, :STREAM, nil)
+      attempt_v6 = local_addr_list.any? {|local_ai| local_ai.afamily == Socket::AF_INET6 }
+      attempt_v4 = local_addr_list.any? {|local_ai| local_ai.afamily == Socket::AF_INET }
+    end
+
+    queue = Queue.new
+
+    if attempt_v6
+      getaddrinfo_v6_th = start_getaddrinfo_v6(host, port, queue, resolv_timeout)
+    end
+
+    if attempt_v4
+      getaddrinfo_v4_th = start_getaddrinfo_v4(host, port, queue, resolv_timeout)
+    end
+
+    if connect_timeout
+      timeout_th = Thread.new do
+        sleep(connect_timeout)
+        queue.close()
+      end
     end
 
-    Addrinfo.foreach(host, port, nil, :STREAM, timeout: resolv_timeout) {|ai|
-      if local_addr_list
-        local_addr = local_addr_list.find {|local_ai| local_ai.afamily == ai.afamily }
-        next unless local_addr
+    socket_list = []
+    getaddrinfo_v6_done = attempt_v6 ? false : true
+    getaddrinfo_v4_done = attempt_v4 ? false : true
+    while ai = queue.pop do
+      break if queue.closed? # timed out
+      case ai
+      when :getaddrinfo_v6_done
+        getaddrinfo_v6_done = true
+      when :getaddrinfo_v4_done
+        getaddrinfo_v4_done = true
       else
         local_addr = nil
+        if local_addr_list
+          local_addr = local_addr_list.find {|local_ai| local_ai.afamily == ai.afamily }
+          next unless local_addr
+        end
+        begin
+          sock = Socket.new(ai.pfamily, ai.socktype, ai.protocol)
+          sock.ipv6only! if ai.ipv6?
+          sock.bind(local_addr) if local_addr
+          case sock.connect_nonblock(ai, exception: false)
+          when 0
+            ret = sock
+            break
+          when :wait_writable
+            socket_list.push(sock)
+            # 250ms is the recommended value for the connection attempt delay in RFC8305
+            _, writable_sockets, = IO.select(nil, socket_list, nil, 0.25)
```

Comment:
:thinking: What about to make 0.25 constant?

Reply:
That's good. I'll make those values constants.